pravega-client-examples/src/main/java/io/pravega/example/streamprocessing/README.md

```diff
@@ -0,0 +1,173 @@
+
+- [EventGenerator](EventGenerator.java):
+  This application generates new events every 1 second
+  and writes them to a Pravega stream (referred to as stream1).
+
+- [ExactlyOnceMultithreadedProcessor](ExactlyOnceMultithreadedProcessor.java):
+  This application continuously reads events from stream1, performs a stateless computation
+  to generate output events, and writes the output event to another
+  Pravega stream (referred to as stream2).
+  It uses the exactly-once algorithms described below.
+
+- [EventDebugSink](EventDebugSink.java):
+  This application reads events from stream2 and displays them
+  on the console.
+
+
+# How to Run
+
+- Enable INFO level logging by editing the file [logback.xml](../../../../../resources/logback.xml).
+  Update it to include:
+  ```
+  <root level="INFO">
+  ```
+
+-  Start the event generator.
+   ```
+   cd pravega-samples
+   ./gradlew pravega-client-examples:startEventGenerator
+   ```
+
+   All parameters are provided as environment variables.
+   You can either set them in your shell (`export PRAVEGA_SCOPE=examples`) or use the below syntax.
+
+   If you are using a non-local Pravega instance, specify the controller as follows:
+   ```
+   PRAVEGA_CONTROLLER=tcp://pravega.example.com:9090 ./gradlew pravega-client-examples:startEventGenerator
+   ```
+
+   Multiple parameters can be specified as follows.
+   ```
+   PRAVEGA_SCOPE=examples PRAVEGA_CONTROLLER=tcp://localhost:9090 ./gradlew pravega-client-examples:startEventGenerator
+   ```
+
+   See [Parameters.java](Parameters.java) for available parameters.
+
+- In another window, start the stream processor:
+  ```
+  ./gradlew pravega-client-examples:startExactlyOnceMultithreadedProcessor
+  ```
+
+- In another window, start the event debug sink:
+  ```
+  ./gradlew pravega-client-examples:startEventDebugSink
+  ```
+
+- Note: The [ExactlyOnceMultithreadedProcessor](ExactlyOnceMultithreadedProcessor.java)
+  will automatically restart from the latest checkpoint.
+  However, if Pravega streams are truncated or deleted, or checkpoint files in
+  /tmp/checkpoints are deleted or otherwise bad, you may need to start over from
+  a clean system. To do so, follow these steps:
+    - Stop the event generator, stream processor, and event debug sink.
+    - Delete the contents of /tmp/checkpoints.
+    - Use a new scope (PRAVEGA_SCOPE) or streams (PRAVEGA_STREAM_1 and PRAVEGA_STREAM_2).
+
+
+# Achieving At-Least-Once Semantics
+
+The current position in a Pravega stream is defined by a stream cut.
+A stream cut is essentially a mapping from segment numbers to byte offsets
+within that segment.
+However, because the number of concurrent segments can dynamically
+increase and decrease for a single stream, it is not possible for
+a Pravega application to maintain this mapping by itself.
+To achieve at-least-once semantics with Pravega, use the algorithms in this section.
+
+In the algorithms below, one process in a distributed system will be designated
```

Comment:
Given that this is an example showing how to work without Flink, I see no reason this code should copy Flink's pattern. For applications that don't have the requirement which Flink does that they be able to rewind the group arbitrarily, it is much easier to eliminate the master process and just rely on automatic checkpoints. The recovery algorithm is simple, upon a checkpoint a worker commits their transaction, and stores their position. Upon failure, readerOffline is called passing the position.

Reply:
The latest commit has at-least-once semantics and does indeed use only automatic checkpoints. It does not use transactions.